lib/mongo/collection/view/iterable.rb

```diff
@@ -35,24 +35,21 @@ module Iterable
         #
         # @yieldparam [ Hash ] Each matching document.
         def each
-          @cursor = nil
           session = client.send(:get_session, @options)
-          if QueryCache.enabled?
-            unless @cursor = cached_cursor
-              @cursor = select_cursor(session)
-              QueryCache.cache_table[cache_key] = @cursor
-            end
+          @cursor = select_cursor(session)
+
+          if QueryCache.enabled? && @cursor.is_a?(Mongo::CachingCursor)
+            QueryCache.cache_table[cache_key] = @cursor
             range = limit || nil
-          else
-            @cursor = select_cursor(session)
           end
+
           if block_given?
-            if !range
-              @cursor.each do |doc|
+            if range
+              @cursor.to_a[0...range].each do |doc|
```

Comment:
I believe I had to figure out what this does exactly previously, if this code stays around can some comments be added for this line and maybe other lines in this method explaining what the logic is?

Reply:
Agreed. I made a comment on RUBY-2367 to add an explanation here if this code still exists.